### PR TITLE
[Bug] Fix NPE problem when process the event if application was cleared already

### PR DESCRIPTION
**Pull request by colinmjj**
### What changes were proposed in this pull request?
There will be NPE problem when process the event if application was cleared already

### Why are the changes needed?
Fix a critical bug which cause resource leak. 

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
UT is added

### CODE REVIEW

File: server/src/main/java/org/apache/uniffle/server/ShuffleFlushManager.java [hunk 1]
```diff
@@ -285,7 +292,9 @@ void processPendingEvents() throws Exception {
           pendingEventTimeoutSec, event.getEvent());
       return;
     }
-    if (storage.canWrite()) {
+    // storage maybe null if the application cache was cleared already
+    // add event to flush queue, and it will be released
+    if (storage == null || storage.canWrite()) {
```
Comment (jerqi) [inline server/src/main/java/org/apache/uniffle/server/ShuffleFlushManager.java:297]: when storage equals `null`, why do we need add it to the FlushQueue?

File: server/src/main/java/org/apache/uniffle/server/ShuffleFlushManager.java [hunk 2]
```diff
@@ -135,59 +139,62 @@ public void addToFlushQueue(ShuffleDataFlushEvent event) {
...
-            event.getEndPartition(),
-            storageBasePaths,
-            shuffleServerId,
-            hadoopConf,
-            storageDataReplica));
+      Storage storage = storageManager.selectStorage(event);
+      // storage info maybe null if the application cache was cleared already
+      if (storage != null) {
+        if (!storage.canWrite()) {
+          addPendingEvents(event);
```
Comment (jerqi) [inline server/src/main/java/org/apache/uniffle/server/ShuffleFlushManager.java:150]: Why do we put the code into 
```
try {
} catch {
} finally {
}
```
We release memory in the finally block, when we add the event to PendingEvents, we shouldn't release the event memory.

File: server/src/main/java/org/apache/uniffle/server/ShuffleFlushManager.java (hunk 1 above)
Comment (colinmjj) [reply server/src/main/java/org/apache/uniffle/server/ShuffleFlushManager.java:297]: Event should be discarded and memory should be updated, these actions can't be finished here

File: server/src/main/java/org/apache/uniffle/server/ShuffleFlushManager.java (hunk 2 above)
Comment (colinmjj) [reply server/src/main/java/org/apache/uniffle/server/ShuffleFlushManager.java:150]: right, it should be out of the try block.

File: server/src/main/java/org/apache/uniffle/server/ShuffleFlushManager.java (hunk 1 above)
Comment (jerqi) [reply server/src/main/java/org/apache/uniffle/server/ShuffleFlushManager.java:297]: Could we change the from
```
 if (System.currentTimeMillis() - event.getCreateTimeStamp() > pendingEventTimeoutSec * 1000L) {
```
to 
```
 if (System.currentTimeMillis() - event.getCreateTimeStamp() > pendingEventTimeoutSec * 1000L || storage == null) {
```
?
It seems better.

Comment (jerqi) [reply server/src/main/java/org/apache/uniffle/server/ShuffleFlushManager.java:297]: Or  we add the code
```
    if (storage == null) {
      if (shuffleServer != null) {
        shuffleServer.getShuffleBufferManager().releaseMemory(
            event.getEvent().getSize(), true, false);
      }
      LOG.error("storage is empty , the event {} is dropped",  event.getEvent());
      return;
    }
```

Comment (colinmjj) [reply server/src/main/java/org/apache/uniffle/server/ShuffleFlushManager.java:297]: ok, I'll update

**Review by jerqi — APPROVED** [review]
LGTM

**Pull request merged**